controllers/clusterresources/opensearchegressrules_controller.go

```diff
@@ -0,0 +1,213 @@
+//+kubebuilder:rbac:groups=clusterresources.instaclustr.com,resources=opensearchegressrules/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the OpenSearchEgressRules object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.4/pkg/reconcile
+func (r *OpenSearchEgressRulesReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	l := log.FromContext(ctx)
+
+	rule := &clusterresourcesv1beta1.OpenSearchEgressRules{}
+	err := r.Client.Get(ctx, types.NamespacedName{
+		Namespace: req.Namespace,
+		Name:      req.Name,
+	}, rule)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return models.ExitReconcile, nil
+		}
+
+		l.Error(err, "Unable to fetch OpenSearch Egress Rules resource")
+
+		return models.ReconcileRequeue, err
+	}
+
+	// It`s handling resource deletion
+	if rule.DeletionTimestamp != nil {
+		err = r.handleDelete(ctx, l, rule)
+		if err != nil {
+			return models.ReconcileRequeue, err
+		}
+
+		return models.ExitReconcile, nil
+	}
+
+	// It`s handling resource creation
+	if rule.Status.ID == "" {
+		err = r.handleCreate(ctx, l, rule)
+		if err != nil {
+			return models.ReconcileRequeue, nil
+		}
+
+		return models.ExitReconcile, nil
+	}
+
+	return models.ExitReconcile, nil
+}
+
+func (r *OpenSearchEgressRulesReconciler) handleCreate(ctx context.Context, l logr.Logger, rule *clusterresourcesv1beta1.OpenSearchEgressRules) error {
+	if rule.Status.ID == "" {
+		rule.Status.ID = fmt.Sprintf("%s~%s~%s", rule.Spec.ClusterID, rule.Spec.Source, rule.Spec.OpenSearchBindingID)
+	}
+
+	_, err := r.API.GetOpenSearchEgressRule(rule.Status.ID)
+	if !errors.Is(err, instaclustr.NotFound) && err != nil {
+		l.Error(err, "failed to get OpenSearch Egress Rule resource from Instaclustr")
+		r.EventRecorder.Eventf(rule, models.Warning, models.CreationFailed,
+			"Failed to get OpenSearch Egress Rule from Instaclustr. Reason: %v", err,
+		)
+
+		return err
+	}
+
+	if errors.Is(err, instaclustr.NotFound) {
+		err = r.API.CreateOpenSearchEgressRules(rule)
+		if err != nil {
+			l.Error(err, "failed to create OpenSearch Egress Rule resource on Instaclustr")
+			r.EventRecorder.Eventf(rule, models.Warning, models.CreationFailed,
+				"Failed to create OpenSearch Egress Rule on Instaclustr. Reason: %v", err,
+			)
+
+			return err
+		}
+	}
+
+	patch := rule.NewPatch()
+
+	// This hack was added because the Instaclustr API returns an egress rule id with a null value, but it should be equal to the pattern {clusterId}~{source}~{bindingId}
+	// This code could be removed after the Instaclustr API team fixes this bug
```

Comment:
Remove this comment or move it in the right place

Reply:
fixed